app/lib/shared/tool_env.dart

```diff
@@ -28,17 +32,32 @@ class ToolEnvRef {
 
   Future release() async {
     _active--;
-    if (_started >= _maxCount && _active == 0) {
-      await _pubCacheDir.delete(recursive: true);
+    if (_active == 0) {
+      // Delete directory if the instance is no longer active or it reached the
+      // maximum threshold.
+      if (_started >= _maxCount || _current != this) {
+        await _pubCacheDir.delete(recursive: true);
+      }
     }
   }
 }
 
-Future<ToolEnvRef> createToolEnvRef() async {
+/// Gets a currently available [ToolEnvRef] if it is used less than the
+/// configured threshold (_maxCount, currently 100). If it it has already
+/// reached the amount, a new cache dir and environment will be created.
+Future<ToolEnvRef> getOrCreateToolEnvRef() async {
   if (_current != null && _current._started < _maxCount) {
     _current._aquire();
     return _current;
   }
+  if (_ongoing == null) {
+    _ongoing = new Completer();
+  } else {
+    // There may be a race condition, it is simpler to just call the method
+    // again after the calculation is done.
+    await _ongoing.future;
+    return await getOrCreateToolEnvRef();
```

Comment:
Unclear how well Dart optimizes this tail call. I recommend using a while loop here instead. Right now you have a thundering herd effect where everything wakes up and races whenever a single work item completes int he queue.

A simple way to do a fair queue is to create a completer. If _ongoing is null, we set it to our future. Otherwise remember the value of _ongoing and replace it with our own future, and then we await that future. We awake and do the work. Then we complete our future. If _ongoing is our future, we set it to null (we were last).